infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesDeployments.java

```diff
@@ -388,19 +392,46 @@ public void onClose(KubernetesClientException cause) {
   }
 
   private void handleStartingPodStatus(CompletableFuture<Void> podRunningFuture, Pod pod) {
-    if (POD_STATUS_PHASE_RUNNING.equals(pod.getStatus().getPhase())) {
-      podRunningFuture.complete(null);
+    PodStatus status = pod.getStatus();
+    String podPhase = status.getPhase();
+    if (POD_STATUS_PHASE_RUNNING.equals(podPhase)) {
+      // check that all the containers are ready...
+      Map<String, String> terminatedContainers = new HashMap<>();
+
+      for (ContainerStatus cs : status.getContainerStatuses()) {
+        ContainerStateTerminated terminated = cs.getState().getTerminated();
+        if (terminated != null) {
+          terminatedContainers.put(
+              cs.getName(),
+              "reason = " + terminated.getReason() + ", message = " + terminated.getMessage());
+        }
+      }
+
+      if (terminatedContainers.isEmpty()) {
```

Comment:
Is it enough to check that the terminated containers list is empty? 
Should not we check every container is ready/running (containerStatus.getReady())?

Reply:
The handler is run on every event from the pod. I therefore only explicitly check for "unrecoverable" error conditions to interrupt the startup, not success conditions. The success is kinda implied by there being no errors. The thinking was that if there are no terminations and the pod is declared running, it is kinda implied that the containers are ready. But maybe not and more experimentation is needed. I'll try that out.